test/distributed/org/apache/cassandra/distributed/test/BootstrapTest.java

```diff
@@ -18,85 +18,241 @@
-        Map<Integer, Long> naturally = null;
-        try (Cluster cluster = builder.withNodes(originalNodeCount).start())
+
+        try (Cluster cluster = builder().withNodes(originalNodeCount)
+                                        .withTokenSupplier(TokenSupplier.evenlyDistributedTokens(expandedNodeCount))
+                                        .withNodeIdTopology(NetworkTopology.singleDcNetworkTopology(expandedNodeCount, "dc0", "rack0"))
+                                        .withConfig(config -> config.with(NETWORK, GOSSIP))
+                                        .start())
         {
-            populate(cluster);
+            populate(cluster,0, 100);
 
             IInstanceConfig config = cluster.newInstanceConfig();
-            config.set("auto_bootstrap", true);
-
-            cluster.bootstrap(config).startup();
+            IInvokableInstance newInstance = cluster.initialize(config);
+            withJoinRing(false, () -> newInstance.startup(cluster));
 
-            cluster.stream().forEach(instance -> {
-                instance.nodetool("cleanup", KEYSPACE, "tbl");
-            });
+            cluster.run(statusToBootstrap(newInstance));
+            cluster.run(pullSchemaFrom(cluster.get(1)), newInstance.config().num());
+            cluster.run(bootstrap(), newInstance.config().num());
 
-            withBootstrap = count(cluster);
+            for (Map.Entry<Integer, Long> e : count(cluster).entrySet())
+                Assert.assertEquals(e.getValue().longValue(), 100L);
         }
+    }
 
-        builder = builder.withNodes(expandedNodeCount)
-                         .withTokenSupplier(TokenSupplier.evenlyDistributedTokens(expandedNodeCount))
-                         .withConfig(config -> config.with(NETWORK, GOSSIP));
+    @Test
+    public void testPendingWrites() throws Throwable
+    {
+        int originalNodeCount = 2;
+        int expandedNodeCount = originalNodeCount + 1;
 
-        try (ICluster cluster = builder.start())
+        try (Cluster cluster = builder().withNodes(originalNodeCount)
+                                        .withTokenSupplier(TokenSupplier.evenlyDistributedTokens(expandedNodeCount))
+                                        .withNodeIdTopology(NetworkTopology.singleDcNetworkTopology(expandedNodeCount, "dc0", "rack0"))
+                                        .withConfig(config -> config.with(NETWORK, GOSSIP))
+                                        .start())
         {
-            populate(cluster);
-            naturally = count(cluster);
+            populate(cluster, 0, 100);
+            IInstanceConfig config = cluster.newInstanceConfig();
+            IInvokableInstance newInstance = cluster.initialize(config);
+            withJoinRing(false, () -> newInstance.startup(cluster));
+
+            cluster.run(statusToBootstrap(newInstance));
+            cluster.run(bootstrap(false, Duration.ofSeconds(60), Duration.ofSeconds(60)), newInstance.config().num());
```

Comment:
been running in a debugger and see that this doesn't properly update the bootstrap state, for this reason join thinks a bootstrap never happened and rebootstraps everything; was also seeing a few other subtle difference than `org.apache.cassandra.service.StorageService#joinTokenRing`.

Rather than try to do this, why not allow startup to bootstrap/join or rely on join to bootstrap (when cassandra.join_ring=false)?  This would actually match the operator flow so the test would touch the normal code paths.

I took a stab at it in https://github.com/dcapwell/cassandra/commit/935f78b101484f6dfee473fd5375a31761f02b39 where I run join in the background, and block streaming (to make sure test wasn't flaky); this seems to mostly match the intent (while bootstrapping there are writes).

Reply:
I've added the missing setting bootstrap state.

Since the intent of this specific test is not to test `joinTokenRing`, but rather to test side-effects of joining: pending ranges, I decided to opt-out for a smaller bootstrap action. In general, the intention of this patch is to be able to reproduce arbitrary cluster states. For example, a state when streaming has finished, but node hasn't yet joined the ring. By opting to `auto_bootstrap`, we're losing this level of control. 

I understand there is more than one way to implement this, and using Byteman can be one of the ways, but in this particular case it's not the one I think is approprate. 

To reiterate, the purpose of this whole patch is exactly to be able to produce such states, and I think your Byteman example shows that the patch has achieved this. If you like, I can include your test as well; however, I do not think it's necessary, since we now have an easier way to do the same.